File: js/medium/checkBrackets.js
function checkBrackets(s) {
  if (!s.trim().length) return false;
  var pairs = {
    open: ['(', '[', '{'],
    close: [')', ']', '}'],
  };
  var stack = [];
  for (var i = 0; i < s.length; i++) {
    var item = s[i];
    if (pairs.open.includes(item)) {
      stack.push(item);
    } else if (pairs.close.includes(item)) {
      var closeIdx = pairs.close.indexOf(item);
      var openBracket = pairs.open[closeIdx];
      if (openBracket === stack[stack.length - 1]) {
        stack.splice(-1, 1);
      } else {
        stack.push(item);
      }
    }
  }
  return stack.length === 0 ? true : false;
}
console.log(checkBrackets('(){}[]')); //  true
console.log(checkBrackets('([{}])')); //  true
console.log(checkBrackets('([{}{}{}])')); //  true
console.log(checkBrackets('([{}{}{}]')); //  false
console.log(checkBrackets('[{}]]]')); //  false
console.log(checkBrackets(' ')); //  false
// O(n)

File: dart/medium/checkBrackets.dart
class CheckBrackets {
  static bool result(String s) {
    if (s.trim().isEmpty) {
      return false;
    }
    List<String> open = ['(', '[', '{'];
    List<String> close = [')', ']', '}'];
    List<String> stack = [];

    for (int i = 0; i < s.length; i++) {
      final String item = s[i];
      if (open.contains(item)) {
        stack.add(item);
      } else if (close.contains(item)) {
        final int openPos = close.indexOf(item);
        if (stack.length > 0 && open[openPos] == stack[stack.length - 1]) {
          stack.removeLast();
        } else {
          return false;
        }
      }
    }
    return stack.isEmpty ? true : false;
  }
}
